File: modules/slicot/src/c/MA01AD.c
/* Translated by Nelson f2c (version 20170901).
   You must link the resulting object file with the libraries:
    -lnlsf2c -lm   (in that order)
*/

#include "nelson_f2c.h"

EXPORTSYMBOL /* Subroutine */ int ma01ad_(xr, xi, yr, yi) doublereal *xr, *xi, *yr, *yi;
{
    /* Builtin functions */
    double sqrt();
    /* Local variables */
    static doublereal s;
    extern doublereal dlapy2_();
    /*     SLICOT RELEASE 5.0. */
    /*     Copyright (c) 2002-2010 NICONET e.V. */
    /*     This program is free software: you can redistribute it and/or */
    /*     modify it under the terms of the GNU General Public License as */
    /*     published by the Free Software Foundation, either version 2 of */
    /*     the License, or (at your option) any later version. */
    /*     This program is distributed in the hope that it will be useful, */
    /*     but WITHOUT ANY WARRANTY; without even the implied warranty of */
    /*     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the */
    /*     GNU General Public License for more details. */
    /*     You should have received a copy of the GNU General Public License */
    /*     along with this program.  If not, see */
    /*     <http://www.gnu.org/licenses/>. */
    /*     PURPOSE */
    /*     To compute the complex square root YR + i*YI of a complex number */
    /*     XR + i*XI  in real arithmetic.  The returned result is so that */
    /*     YR >= 0.0  and  SIGN(YI) = SIGN(XI). */
    /*     ARGUMENTS */
    /*     Input/Output Parameters */
    /*     XR      (input) DOUBLE PRECISION */
    /*     XI      (input) DOUBLE PRECISION */
    /*             These scalars define the real and imaginary part of the */
    /*             complex number of which the square root is sought. */
    /*     YR      (output) DOUBLE PRECISION */
    /*     YI      (output) DOUBLE PRECISION */
    /*             These scalars define the real and imaginary part of the */
    /*             complex square root. */
    /*     METHOD */
    /*     The complex square root YR + i*YI of the complex number XR + i*XI */
    /*     is computed in real arithmetic, taking care to avoid overflow. */
    /*     REFERENCES */
    /*     Adapted from EISPACK subroutine CSROOT. */
    /*     CONTRIBUTOR */
    /*     P. Benner, Universitaet Bremen, Germany, and */
    /*     R. Byers, University of Kansas, Lawrence, USA, */
    /*     Aug. 1998, routine DCROOT. */
    /*     V. Sima, Research Institute for Informatics, Bucharest, Romania, */
    /*     Oct. 1998, SLICOT Library version. */
    /*     REVISIONS */
    /*     - */
    /*     ****************************************************************** */
    /*     .. Parameters .. */
    /*     .. */
    /*     .. Scalar Arguments .. */
    /*     .. */
    /*     .. Local Scalars .. */
    /*     .. */
    /*     .. External Functions .. */
    /*     .. Intrinsic functions .. */
    /*     .. */
    /*     .. Executable Statements .. */
    s = sqrt((dlapy2_(xr, xi) + abs(*xr)) * .5);
    if (*xr >= 0.) {
        *yr = s;
    }
    if (*xi < 0.) {
        s = -s;
    }
    if (*xr <= 0.) {
        *yi = s;
        if (*xr < 0.) {
            *yr = *xi / s * .5;
        }
    } else {
        *yi = *xi / *yr * .5;
    }
    return 0;
    /*     *** Last line of MA01AD *** */
} /* ma01ad_ */

File: modules/slicot/src/fortran/MA01AD.f
      SUBROUTINE MA01AD( XR, XI, YR, YI )
C
C     PURPOSE
C
C     To compute the complex square root YR + i*YI of a complex number
C     XR + i*XI  in real arithmetic.  The returned result is so that
C     YR >= 0.0  and  SIGN(YI) = SIGN(XI).
C
C     ARGUMENTS
C
C     Input/Output Parameters
C
C     XR      (input) DOUBLE PRECISION
C     XI      (input) DOUBLE PRECISION
C             These scalars define the real and imaginary part of the
C             complex number of which the square root is sought.
C
C     YR      (output) DOUBLE PRECISION
C     YI      (output) DOUBLE PRECISION
C             These scalars define the real and imaginary part of the
C             complex square root.
C
C     METHOD
C
C     The complex square root YR + i*YI of the complex number XR + i*XI
C     is computed in real arithmetic, taking care to avoid overflow.
C
C     REFERENCES
C
C     Adapted from EISPACK subroutine CSROOT.
C
C     CONTRIBUTOR
C
C     P. Benner, Universitaet Bremen, Germany, and
C     R. Byers, University of Kansas, Lawrence, USA,
C     Aug. 1998, routine DCROOT.
C     V. Sima, Research Institute for Informatics, Bucharest, Romania,
C     Oct. 1998, SLICOT Library version.
C
C     REVISIONS
C
C     -
C
C     ******************************************************************
C
C     .. Parameters ..
      DOUBLE PRECISION  ZERO, HALF
      PARAMETER         ( ZERO = 0.0D0, HALF = 1.0D0/2.0D0 )
C     ..
C     .. Scalar Arguments ..
      DOUBLE PRECISION  XR, XI, YR, YI
C     ..
C     .. Local Scalars ..
      DOUBLE PRECISION  S
C     ..
C     .. External Functions ..
      DOUBLE PRECISION  DLAPY2
      EXTERNAL          DLAPY2
C
C     .. Intrinsic functions ..
      INTRINSIC         ABS, SQRT
C     ..
C     .. Executable Statements ..
C
      S = SQRT( HALF*( DLAPY2( XR, XI ) + ABS( XR ) ) )
      IF ( XR.GE.ZERO ) YR =  S
      IF ( XI.LT.ZERO ) S  = -S
      IF ( XR.LE.ZERO ) THEN
         YI =  S
         IF ( XR.LT.ZERO ) YR =  HALF*( XI/S )
      ELSE
         YI =  HALF*( XI/YR )
      END IF
C
      RETURN
C     *** Last line of MA01AD ***
      END
